x-pack/plugins/infra/public/containers/metrics/with_metrics_time.tsx

```diff
@@ -4,89 +4,152 @@
-export const withMetricsTime = connect(
-  (state: State) => ({
-    currentTimeRange: metricTimeSelectors.selectRangeTime(state),
-    isAutoReloading: metricTimeSelectors.selectIsAutoReloading(state),
-    urlState: selectTimeUrlState(state),
-  }),
-  bindPlainActionCreators({
-    setRangeTime: metricTimeActions.setRangeTime,
-    startMetricsAutoReload: metricTimeActions.startMetricsAutoReload,
-    stopMetricsAutoReload: metricTimeActions.stopMetricsAutoReload,
-  })
-);
+interface MetricsTimeState {
+  timeRange: InfraTimerangeInput;
+  setTimeRange: (timeRange: InfraTimerangeInput) => void;
+  refreshInterval: number;
+  setRefreshInterval: (refreshInterval: number) => void;
+  isAutoReloading: boolean;
+  setAutoReload: (isAutoReloading: boolean) => void;
+}
+
+export const useMetricsTime = () => {
+  const [isAutoReloading, setAutoReload] = useState(false);
+  const [refreshInterval, setRefreshInterval] = useState(5000);
+  const [timeRange, setTimeRange] = useState({
+    from: moment()
+      .subtract(1, 'hour')
+      .valueOf(),
+    to: moment().valueOf(),
+    interval: '>=1m',
+  });
+
+  useInterval(
+    () => {
+      const range = timeRange.to - timeRange.from;
+      setTimeRange({
+        from: moment()
+          .subtract(range, 'ms')
+          .valueOf(),
+        to: moment().valueOf(),
+        interval: '>=1m',
+      });
+    },
+    isAutoReloading ? refreshInterval : null
+  );
+
+  useEffect(
+    () => {
+      if (isAutoReloading) {
+        const range = timeRange.to - timeRange.from;
+        setTimeRange({
+          from: moment()
+            .subtract(range, 'ms')
+            .valueOf(),
+          to: moment().valueOf(),
+          interval: '>=1m',
+        });
+      }
+    },
+    [isAutoReloading]
+  );
 
-export const WithMetricsTime = asChildFunctionRenderer(withMetricsTime, {
-  onCleanup: ({ stopMetricsAutoReload }) => stopMetricsAutoReload(),
-});
+  return {
+    timeRange,
+    setTimeRange,
+    refreshInterval,
+    setRefreshInterval,
+    isAutoReloading,
+    setAutoReload,
+  };
+};
+
+export const MetricsTimeContainer = createContainer(useMetricsTime);
+
+interface WithMetricsTimeProps {
+  children: (args: MetricsTimeState) => React.ReactElement;
+}
+export const WithMetricsTime: React.FunctionComponent<WithMetricsTimeProps> = ({
+  children,
+}: WithMetricsTimeProps) => {
+  const metricsTimeState = useContext(MetricsTimeContainer.Context);
+  return children({ ...metricsTimeState });
+};
 
 /**
  * Url State
  */
 
-interface MetricTimeUrlState {
-  time?: ReturnType<typeof metricTimeSelectors.selectRangeTime>;
-  autoReload?: ReturnType<typeof metricTimeSelectors.selectIsAutoReloading>;
+interface MetricsTimeUrlState {
+  time?: MetricsTimeState['timeRange'];
+  autoReload?: boolean;
+  refreshInterval?: number;
 }
 
 export const WithMetricsTimeUrlState = () => (
   <WithMetricsTime>
-    {({ setRangeTime, startMetricsAutoReload, stopMetricsAutoReload, urlState }) => (
+    {({
```

Comment:
This could just `useContext(MetricsTimeContainer.Context)` directly instead of the container component, couldn't it?

Reply:
It could, the reason I kept the container pattern was due to the index page which is also using the `WithMetricsTime` component. 

On the index page I've hoisted `<MetricsTimeContainer.Provider>` as high as I can, but due to `Provider`s not accepting functions as children and the `MetricDetail` page being a class component not a functional component I can't do a `useContext()` call conveniently in that `render` body (whilst also being after the `<MetricsTimeContainer.Provider>`). I didn't want to hoist the `Provider` to the routes level, as that muddies up the tidy routes representation. 

I could of course remove the use from `WithMetricsTimeUrlState` but because the `MetricDetail` component was going to be using it anyway, I thought it made most sense to use it between both.

Reply:
Oh, right. Using the container makes sense then. Thanks for the explanation.